src/tfr_control/include/tfr_control/controller.h

```diff
@@ -0,0 +1,57 @@
+/**
+ * controller.h
+ * 
+ * This class is in charge of handling the physical hardware interface with
+ * the robot itself, and is started by the controller_launcher node.
+ */
+#ifndef CONTROLLER_H
+#define CONTROLLER_H
+
+#include <ros/ros.h>
+#include <std_msgs/Float64.h>
+#include <hardware_interface/joint_command_interface.h>
+#include <hardware_interface/joint_state_interface.h>
+#include <hardware_interface/robot_hw.h>
+
+namespace tfr_control {
+    enum Actuator { kLeftTread, kRightTread, kBin, kTurntable, kLowerArm,
+                    kUpperArm, kScoop };
+
+    /**
+     * The class registered with controller_manager that handles the interface
+     * with all hardware components on the robot. See controller_launcher.cpp
+     * for the usage of this class.
+     * */
+    class Controller : public hardware_interface::RobotHW
+    {
+    public:
+        Controller();
+        void read();
+        void write();
+    private:
+        hardware_interface::JointStateInterface joint_state_interface;
+        hardware_interface::EffortJointInterface joint_effort_interface;
+        
+        /**
+         * Seven actuators/motors to store data for:
+         *  - Left + right tread
+         *  - Bin actuator
+         *  - Turntable
+         *  - Lower arm actuator
+         *  - Upper arm actuator
+         *  - Scoop actuator
+         * */
+        void register_joint(std::string joint, Actuator actuator);
+
+        // Populated by controller_manager, read in by the write() method
+        double command_values[7];
```

Comment:
What could they be initialized with? Maybe I'm wrong here, but don't they have to initialized by the controller_manager with their real physical values? If so, then any values given to them will not be any better than whatever random stuff was in memory.

Maybe we could create a variable that tracks if they have been initialized with write, that throws an error if they have not been written to yet?

Reply:
Thats a little complicated. The brace will initialize them to 0. If we want a different value we could do:

Arr my Art[]={-1}
Its a one line fix,and initializing arrays is a best practice in c++11.

Keep simple things simple

Reply:
I'm fine with the change. I'm just not sure it actually helps (zero doesn't seem any better than random stuff in memory if both are wrong).

So I think the best thing to do would be initializing them to something that is obviously wrong so the program doesn't "accidentally work" sometimes and fail at others randomly.

Reply:
I agree with Levi.  Initializing variables to 0 or some other default value is best practice because sometimes people assume incorrectly that 0 *is* the default value, and use them as though that were the case.  Here, *any* usage of the values before they've been initialized by the ROS framework is unacceptable.

I would recommend adding an `initialized` flag and then putting `assert` statements in `read` and `write` to check the flag.  Or we could throw exceptions in those functions instead, since ROS makes heavy use of exceptions anyway.

Reply:
I think I'd prefer initializing with zero vs any other value as that seems more likely to result in less random hardware movement (if something were to happen perchance). Adding a flag also seems definitely doable, I'll experiment with `assert` vs exceptions vs something else (if it exists) and get back to you on that.

Reply:
@RyanBerge @LinuxKid2 @AdamuKaapan  If we are going to make any sort of assumption about that, then let's just leave it. I did not intend for this to become any sort of contested issue. My thing is if I see an uninitialized variable in code I have been assigned to review. I'll mark it, it's on my list of things that should be done. 

I think this exception thing is excessive, it should just be  a quick fix and move on.

Reply:
No worries. Initializing it I don't think will be an issue anyways, I'll go ahead and do that, I don't think we need to worry about checking it.